### PR TITLE
Fix copying Message in pure-python.

### LINKED ISSUE #14: Regression in 4.3 Python impl: NoneType object is not subscriptable
Seen running the zope.app.apidoc tests, but probably also shows up with zope.app.preference. The same tests pass with 4.2 and under CPython:
```
      File "//pypy/site-packages/zope/app/apidoc/browser/preference.py", line 40, in __init__
        super(ApidocEditPreferenceGroup, self).__init__(context, request)
      File "//pypy/site-packages/zope/app/preference/browser.py", line 84, in __init__
        self.label = Message(pref_msg, mapping={u'name': name})
      File "//pypy/site-packages/zope/i18nmessageid/message.py", line 41, in __new__
        default = ustr.default[:] if default is _marker else default
    TypeError: 'NoneType' object is not subscriptable
```

### PR DESCRIPTION
**Pull request by jamadden**
Do it like the C extension does, by actually copying from the other object and then setting attributes if needed.

Also drop old C support code for Python < 2.6

Fixes #14

### CODE REVIEW

File: src/zope/i18nmessageid/message.py [hunk 1]
```diff
@@ -37,37 +37,37 @@ def __new__(cls, ustr, domain=_marker, default=_marker, mapping=_marker,
...
-            default_plural = (
-                ustr.default_plural[:] if default_plural is _marker else
-                default_plural)
-            number = ustr.number if number is _marker else number
-            ustr = six.text_type(ustr)
+            self.domain = ustr.domain[:] if ustr.domain else None
+            self.default = ustr.default[:] if ustr.default else None
+            self.mapping = ustr.mapping.copy() if ustr.mapping is not None else None
+            self.msgid_plural = ustr.msgid_plural[:] if ustr.msgid_plural else None
+            self.default_plural = ustr.default_plural[:] if ustr.default_plural else None
```
Comment (mgedmin) [inline src/zope/i18nmessageid/message.py]: Why are we even trying to slice strings anyway?  `a_string[:] is a_string` in CPython AFAICT.  Is the C code doing slices?

(Also I'm mildly unhappy about explicitly making a copy of the mapping dictionary and then throwing it away.  I suppose it's a rare occurrence, but still.)

Comment (jamadden) [reply src/zope/i18nmessageid/message.py]: I wondered that too. The C code is *not* slicing strings. Nor does it `copy()` the mapping. Suppose we just remove that here?

Comment (jamadden) [reply src/zope/i18nmessageid/message.py]: No tests broke if that stops happening. I removed it and added tests to confirm that Python and C  are handling the objects the same way.

**Review by mgedmin — APPROVED** [review]
LGTM

(but the change that introduced the bug also looked good to me so caveat emptor)

**Comment by jamadden** [comment]
Thanks! Let's roll the dice. Can we get a PyPI release, please? (Or access for `jamadden`)

**Comment by jamadden** [comment]
I started the release process and uploaded the git tags before I realised I didn't have PyPI rights. So the 4.3.1 wheels are showing up on PyPI now thanks to the CI services, but not a sdist (.tar.gz).

**Comment by tseaver** [comment]
I just added you as an owner of the PyPI project.

**Comment by jamadden** [comment]
Thank you! I've uploaded the corresponding .tar.gz

**Comment by mgedmin** [comment]
For packages where CI uploads wheels perhaps we should also upload sdists (using `twine upload --skip-missing`), just in case.

**Pull request merged**